server-utils/src/suspendable_stream.rs

```diff
@@ -40,30 +40,34 @@ impl<S, I> Stream for SuspendableStream<S>
 	type Error = ();
 
 	fn poll(&mut self) -> Result<Async<Option<Self::Item>>, ()> {
-		if let Some(mut timeout) = self.timeout.take() {
-			match timeout.poll() {
-				Ok(Async::Ready(_)) => {}
-				Ok(Async::NotReady) => {
-					self.timeout = Some(timeout);
-					return Ok(Async::NotReady);
+		loop {
+			if let Some(mut timeout) = self.timeout.take() {
+				match timeout.poll() {
+					Ok(Async::Ready(_)) => {}
+					Ok(Async::NotReady) => {
+						self.timeout = Some(timeout);
+						return Ok(Async::NotReady);
+					}
+					Err(err) => {
+						warn!("Timeout error {:?}", err);
+						task::current().notify();
+						return Ok(Async::NotReady);
+					}
 				}
```

Comment:
You should break out of the loop in case `timeout = None`!

Reply:
If we do that, we would never call `self.stream.poll()`.

Reply:
Right, apologies. I misread the code and though that only `self.timeout` is inside the `loop`